bin/cli.js

```diff
@@ -1,17 +1,245 @@
+};
+
+const getReporter = () => {
+  const browsers = [];
+  const reporter = function(base) {
+    base(this);
+    this.onBrowserLog = (browser, log) => {
+      if (typeof(log) !== 'string') log = JSON.stringify(log);
+      if (browsers.indexOf(browser) === -1) {
+        browsers.push(browser);
+        console.log(`\n${browser}:`);
+      }
+      console.log(log.slice(1, log.length - 1));
+    };
+  };
+
+  reporter.$inject = ['baseReporterDecorator'];
+  return reporter;
+};
+
+const getBrowserConfig = (conf) => {
+  const config = {
+    preprocessors: {},
+    files: [],
+    plugins: [
+      'karma-webpack',
+      { 'reporter:meta': ['type', getReporter()] }
+    ],
+    reporters: [ 'meta' ],
+    basePath: process.env.PWD,
+    port: conf.browser.port,
+    autoWatch: false,
+    singleRun: true,
+    concurrency: 1
+  };
+
+  const adapter = path.resolve('./build/adapter.js');
+  config.files.push(adapter);
+  config.preprocessors[adapter] = ['webpack'];
+
+  const nodePackages = [ 'fs', 'child_process' ];
+
+  removeNodePackages(config, ...nodePackages);
+  setKarmaBrowsers(config, ...conf.browser.browsers);
+  setKarmaLogLevel(config, conf.browser.logLevel);
+  return config;
+};
+
+const getEnvironment = (config, program) => {
+  if (program.node || program.browser) {
+    config.environments = [];
+    if (program.node) config.environments.push('node');
+    if (program.browser) config.environments.push('browser');
+  }
+  if (!config.environments || config.environments.length === 0)
+    config.environments = [ 'node' ];
+};
+
+const getConfig = () => {
+  const version = parseFile(path.resolve(__dirname, '../package.json')).version;
+  program.version(version).usage('[options] -- <file ...>');
+  cliOptions.forEach(option => program.option(...option));
+  program.parse(process.argv);
+
+  const config = program.config ? parseFile(program.config) : {};
+  getEnvironment(config, program);
+
+  config.files = merge(config.files, program.args);
+  config.files = loadFiles(config.files);
+
+  config.exclude = merge(config.exclude, program.exclude);
+  config.files = exclude(config.files, config.exclude);
+  config.nodeOnly = merge(config.nodeOnly, program.nodeOnly);
+  config.browserOnly = merge(config.browserOnly, program.browserOnly);
+
+  if (config.environments.indexOf('browser') >= 0) {
+    config.browser = config.browser || {};
+
+    config.browser.browsers = merge(
+      config.browser.browsers, program.browsers);
+    config.browser.logLevel = program.browserLog || config.browser.logLevel;
+    config.browser.port = +program.browserPort || config.browser.port;
+    config.browser = getBrowserConfig(config);
+  }
+  return config;
+};
+
+const runNode = (config, cb) => {
+  metatests.runner.instance.on('finish', (failed) => {
+    console.log('Tests finished. Waiting for unfinished tests after end...\n');
+    const code = failed ? 1 : 0;
+    setTimeout(() => cb(code), 5000);
```

Comment:
Because some tests can send info after `metatests` `finish` event, so we need to wait some time before exit. As for amount, @lundibundi said that this will be enough.

Reply:
As I said in the metatests refactoring there is no way of determining 'tests finish' as some tests may fail after emitting proper end (i.e. 'test after end', 'end after end' etc checks) and 'error' event will be emitted. This is due to the way we currently handle tests itself and I didn't want to add even more changes to that PR in order to fix that as it was already big enough.

As for the 5 seconds, it's just some random number to wait for some tests to maybe emit error as there is no other way to know of such errors other than waiting as of now. Maybe put 3s there to not wait as long for each test run. Also, note that this will apply only to `cli` usage as it manually calls `process.exit`. If we directly run tests with `node test.js` it will print all of the errors as all of the remaining code to execute that can possibly emit errors will keep the event loop alive and therefore you will get all of the errors.

Reply:
Also, I don't think waiting for 5 seconds here accomplishes anything since you are saving the value of the `code`, when you call the `setTimeout`, inside the scope of the function that you pass to the `setTimeout` and it is not changed anywhere else (it is also a `const` value).